test/pytest/test_bridge.py

```diff
@@ -0,0 +1,114 @@
+
+    def send_init(self, version=1, host='test', **kwargs):
+        self.send_json('', command='init', version=version, host=host, **kwargs)
+
+    def send_open(self, channel, payload, **kwargs):
+        self.send_json('', command='open', channel=channel, payload=payload, **kwargs)
+
+    def send_done(self, channel, **kwargs):
+        self.send_json('', command='done', channel=channel, **kwargs)
+
+    def send_close(self, channel, **kwargs):
+        self.send_json('', command='close', channel=channel, **kwargs)
+
+    def send_ping(self, **kwargs):
+        self.send_json('', command='ping', **kwargs)
+
+    def __init__(self, protocol: asyncio.Protocol):
+        self.queue = asyncio.Queue()
+        self.protocol = protocol
+        protocol.connection_made(self)
+
+    def write(self, data: bytes) -> None:
+        # We know that the bridge only ever writes full frames at once, so we
+        # can disassemble them immediately.
+        _, channel, data = data.split(b'\n', 2)
+        self.queue.put_nowait((channel.decode('ascii'), data))
+
+    def close(self) -> None:
+        pass
+
+    async def next_frame(self) -> Tuple[str, bytes]:
+        return await self.queue.get()
+
+    async def next_msg(self, expected_channel) -> Dict[str, Any]:
+        channel, data = await self.next_frame()
+        assert channel == expected_channel
+        return json.loads(data)
+
+    async def assert_data(self, expected_channel: str, expected_data: bytes) -> None:
+        channel, data = await self.next_frame()
+        assert channel == expected_channel
+        assert data == expected_data
+
+    async def assert_msg(self, expected_channel, **kwargs) -> None:
+        msg = await self.next_msg(expected_channel)
+        assert msg == msg | {k.replace('_', '-'): v for k, v in kwargs.items()}
+
+
+class TestBridge(unittest.IsolatedAsyncioTestCase):
+    async def start(self) -> Tuple[Router, MockTransport]:
+        bridge = Router()
+        transport = MockTransport(bridge)
+
+        await transport.assert_msg('', command='init')
+        transport.send_init()
+
+        return bridge, transport
+
+    async def test_echo(self):
+        bridge, transport = await self.start()
+
+        transport.send_open('1', 'echo')
+        await transport.assert_msg('', command='ready', channel='1')
+
+        transport.send_data('1', b'foo')
+        await transport.assert_data('1', b'foo')
+
+        transport.send_ping(channel='1')
+        await transport.assert_msg('', command='pong', channel='1')
+
+        transport.send_done('1')
+        await transport.assert_msg('', command='done', channel='1')
+
+        transport.send_close('1')
+        await transport.assert_msg('', command='close', channel='1')
+
+    async def test_host(self):
+        bridge, transport = await self.start()
+
+        # try to open an null channel, explicitly naming our host
+        transport.send_open('1', 'null', host='test')
```

Comment:
It took me a bit to figure out why/how `test` magic. Would you mind introducing a global constant for this, and using it here and in `send_init()`?

Reply:
Sure.